File: test_filter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@file: test_filter.py
@brief: 一阶低通滤波Python版测试用例
@author: ptmicky
@version: 1.0
@date: 2026-03-12
"""

from first_order_filter import first_order_filter
import numpy as np

def test_filter():
    # 模拟含噪声数据
    raw_data = [20, 25, 18, 22, 21, 24, 19, 23]
    k = 0.8
    last_y = raw_data[0]
    
    # 批量滤波
    filtered_data, last_y = first_order_filter(raw_data, last_y, k)
    
    # 输出结果
    print("=== Python版一阶低通滤波测试结果 ===")
    print(f"原始数据：{raw_data}")
    print(f"滤波后数据（k={k}）：{[round(y, 2) for y in filtered_data]}")

    # 可视化（可选）
    try:
        import matplotlib.pyplot as plt
        plt.figure(figsize=(8, 4))
        plt.plot(raw_data, 'ro-', label='原始数据', linewidth=1.5, markersize=6)
        plt.plot(filtered_data, 'b*-', label='滤波后数据', linewidth=1.5, markersize=6)
        plt.xlabel('采样点')
        plt.ylabel('数值')
        plt.title(f'一阶低通滤波效果（k={k}）')
        plt.grid(True)
        plt.legend()
        plt.savefig("python_filter_demo.png", dpi=100, bbox_inches='tight')
        plt.show()
    except ImportError:
        print("提示：未安装matplotlib，跳过可视化（可执行 pip install matplotlib 安装）")

if __name__ == "__main__":
    test_filter()

File: test_filter.c
/**
 * @file test_filter.c
 * @brief 一阶低通滤波测试用例
 * @author ptmicky
 * @version 1.0
 * @date 2026-03-12
 */

#include "first_order_filter.h"
#include <stdio.h>

int main(void)
{
    // 模拟含噪声的传感器数据
    float raw_data[] = {20.0f, 25.0f, 18.0f, 22.0f, 21.0f, 24.0f, 19.0f, 23.0f};
    uint16_t length = sizeof(raw_data) / sizeof(raw_data[0]);
    float filtered_data[length] = {0}; // 滤波结果数组
    float k = 0.8f; // 滤波系数
    float last_y = raw_data[0]; // 初始化历史值

    // 批量滤波
    batch_filter(raw_data, filtered_data, length, k, &last_y);

    // 输出结果
    printf("=== C版一阶低通滤波测试结果 ===\n");
    printf("原始数据：");
    for (uint16_t i = 0; i < length; i++) {
        printf("%.1f ", raw_data[i]);
    }
    printf("\n滤波后数据（k=%.1f）：", k);
    for (uint16_t i = 0; i < length; i++) {
        printf("%.2f ", filtered_data[i]);
    }
    printf("\n");

    return 0;
}
